File: LeetCode/206. Reverse Linked List.py
# https://leetcode.com/problems/reverse-linked-list/description/

# Оба решения хорошие

# Решение 1
# class Solution:
#     def reverseList(self, head, prev=None):
#         if not head:
#             return prev

#         temp = head.next

#         head.next = prev

#         return self.reverseList(temp, head)


# Решение 2
class Solution:
    def reverseList(self, node):
        if not node:
            return node

        current_node = node
        next_node = node.next

        prev_node = None

        while current_node and next_node:
            temp_next_node = next_node.next

            next_node.next = current_node
            current_node.next = prev_node

            prev_node = current_node

            current_node = next_node

            next_node = temp_next_node

        return current_node


solution = Solution()

File: LeetCode/206. Reverse Linked List.js
// https://leetcode.com/problems/reverse-linked-list/description/

// Оба решения хорошие

// Решение 1
// const reverseList = (node, prev = null) => {
//   if (!node) return prev;

//   const temp = node.next;

//   node.next = prev;

//   return reverseList(temp, node);
// };

// Решение 2
const reverseList = (head) => {
  if (!head) return head;

  let current = head;

  let next = current.next;

  let prev = null;

  while (current && next) {
    const temp = next.next;

    next.next = current;

    current.next = prev;

    prev = current;

    current = next;

    next = temp;
  }

  return current;
};
